rowexec: add batchedInvertedExprEvaluator for inverted index execution

sumeerbhola opened the pull request:
This is for evaluating invertedexpr.SpanExpressions, both for join
queries and single table filtering, using an inverted index.

This PR depends on #48466 which is the first commit. More usage
context can be found in #48019 which contains an invertedJoiner
that uses an earlier version of this evaluator. I am looking for
a review of the structure before adding tests.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/48720)
<!-- Reviewable:end -->

rytaft reviewed (commented):
Structure looks good to me!

> Reviewed 3 of 3 files at r1, 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48720#-:-M7EvAu34vrmRRtbdXQU:b-t7ncn)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @sumeerbhola)

___
*[pkg/sql/rowexec/inverted_expr_evaluator.go, line 105 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48720#-M7EggMFAYwPOSQRwr3s:-M7EggMGBaW2ebiOqSAd:bwqwjd8) ([raw file](https://github.com/cockroachdb/cockroach/blob/4306dcb090fc0bdb224fb49f606cb1b423d6f4aa/pkg/sql/rowexec/inverted_expr_evaluator.go#L105)):*
> ```Go
> 	op invertedexpr.SetOperator
> 	// The index in invertedExprEvaluator.sets
> 	unionSet int
> ```

[nit] `unionSet` seems like a misleading name if it's supposed to be an index... Maybe call this `setIndex` similar to below?
___
*[pkg/sql/rowexec/inverted_expr_evaluator.go, line 121 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48720#-M7Em3Xc0RMJxqlhdLt1:-M7Em3Xc0RMJxqlhdLt2:b3xi9oj) ([raw file](https://github.com/cockroachdb/cockroach/blob/4306dcb090fc0bdb224fb49f606cb1b423d6f4aa/pkg/sql/rowexec/inverted_expr_evaluator.go#L121)):*
> ```Go
> // used -- see batchedInvertedExprEvaluator.
> type invertedExprEvaluator struct {
> 	expr    *invertedexpr.SpanExpression
> ```

Do you actually need to store the original expr? Seems like it's not used after initialization.
___
*[pkg/sql/rowexec/inverted_expr_evaluator.go, line 220 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48720#-M7EkheSDsWMAiokQl6E:-M7EkheSDsWMAiokQl6F:b-6h4bd3) ([raw file](https://github.com/cockroachdb/cockroach/blob/4306dcb090fc0bdb224fb49f606cb1b423d6f4aa/pkg/sql/rowexec/inverted_expr_evaluator.go#L220)):*
> ```Go
> }
> 
> // invertedValRoutingInfo contains the list of exprAndSetIndex pairs that
> ```

invertedValRoutingInfo -> invertedSpanRoutingInfo
___
*[pkg/sql/rowexec/inverted_expr_evaluator.go, line 221 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48720#-M7EkrYU6t9GMgl0J92i:-M7EkrYU6t9GMgl0J92j:b-wkgy04) ([raw file](https://github.com/cockroachdb/cockroach/blob/4306dcb090fc0bdb224fb49f606cb1b423d6f4aa/pkg/sql/rowexec/inverted_expr_evaluator.go#L221)):*
> ```Go
> 
> // invertedValRoutingInfo contains the list of exprAndSetIndex pairs that
> // needs rows from the inverted index span. A []invertedSpanRoutingInfo with
> ```

[nit] needs -> need
___
*[pkg/sql/rowexec/inverted_expr_evaluator.go, line 231 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48720#-M7ElGGp3EBfAs25dAsN:-M7ElGGp3EBfAs25dAsO:b-18lm5i) ([raw file](https://github.com/cockroachdb/cockroach/blob/4306dcb090fc0bdb224fb49f606cb1b423d6f4aa/pkg/sql/rowexec/inverted_expr_evaluator.go#L231)):*
> ```Go
> // batchedInvertedExprEvaluator is for evaluating one or more expressions.
> // The batched evaluator can be reused by calling reset(). In the build phase,
> // append expressions directly to exprs. A nil RPInvertedIndexExpr is
> ```

RPInvertedIndexExpr -> *invertedexpr.SpanExpression
___
*[pkg/sql/rowexec/inverted_expr_evaluator.go, line 265 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48720#-M7En74x2XjP-vdOmRFW:-M7En74y4DHysR929W8F:b-te60cl) ([raw file](https://github.com/cockroachdb/cockroach/blob/4306dcb090fc0bdb224fb49f606cb1b423d6f4aa/pkg/sql/rowexec/inverted_expr_evaluator.go#L265)):*
> ```Go
> //
> // For the c-e span, all the exprAndSetIndexList slices for these spans are
> // unioned since any row in that span needs to be routed to all these
> ```

Not sure I understand what you mean by "the spans are unioned" in this context. Isn't it just that you append the `exprAndSetIndexList` slice from each matching item in `pendingSpans` to the output `exprAndSetIndexList` slice in `fragmentedSpans`? I think you may be overloading the term "union" here....
___
*[pkg/sql/rowexec/inverted_expr_evaluator.go, line 319 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48720#-M7Epcch6Wus7T4ukfKJ:-M7Epcci9n_On8HEdwnE:b-jibl8r) ([raw file](https://github.com/cockroachdb/cockroach/blob/4306dcb090fc0bdb224fb49f606cb1b423d6f4aa/pkg/sql/rowexec/inverted_expr_evaluator.go#L319)):*
> ```Go
> 				b.pendingSpans[i].span.Start = nextStart
> 			}
> 			// All spans in pendingSpans contribute to exprAndOperand.
> ```

exprAndOperand -> exprAndSetIndexList
___
*[pkg/sql/rowexec/inverted_expr_evaluator.go, line 326 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48720#-M7EqUZ0BD5MLkDQWXTK:-M7EqUZ0BD5MLkDQWXTL:b-bz4p9q) ([raw file](https://github.com/cockroachdb/cockroach/blob/4306dcb090fc0bdb224fb49f606cb1b423d6f4aa/pkg/sql/rowexec/inverted_expr_evaluator.go#L326)):*
> ```Go
> 		b.pendingSpans = b.pendingSpans[removeSize:]
> 		if removeSize == 0 {
> 			break
> ```

I think this could use a comment to remind us why you're breaking here.


<!-- Sent from Reviewable.io -->

sumeerbhola committed:
rowexec: add batchedInvertedExprEvaluator for inverted index execution

This is for evaluating invertedexpr.SpanExpressionProto, both for join
queries and single table filtering, using an inverted index.

More usage context can be found in #48019 which contains an invertedJoiner
that uses an earlier version of this evaluator.

Release note: None

sumeerbhola reviewed (commented):
TFTR!

I've addressed the comments and added unit tests, so this should be ready for a full review.
Also, the evaluator now uses `SpanExpressionProto` (which is the first commit, from the other PR)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48720#-:-M7rPBZ87gGsG86fFixU:b-42sfbx)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @rytaft)

___
*[pkg/sql/rowexec/inverted_expr_evaluator.go, line 105 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48720#-M7EggMFAYwPOSQRwr3s:-M7deeRd-z0sLgDKlsn_:bff0mor) ([raw file](https://github.com/cockroachdb/cockroach/blob/4306dcb090fc0bdb224fb49f606cb1b423d6f4aa/pkg/sql/rowexec/inverted_expr_evaluator.go#L105)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] `unionSet` seems like a misleading name if it's supposed to be an index... Maybe call this `setIndex` similar to below?
</blockquote></details>

Renamed to `unionSetIndex`
___
*[pkg/sql/rowexec/inverted_expr_evaluator.go, line 121 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48720#-M7Em3Xc0RMJxqlhdLt1:-M7dfRHn0v-WWnGMjxgf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4306dcb090fc0bdb224fb49f606cb1b423d6f4aa/pkg/sql/rowexec/inverted_expr_evaluator.go#L121)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Do you actually need to store the original expr? Seems like it's not used after initialization.
</blockquote></details>

Done.
___
*[pkg/sql/rowexec/inverted_expr_evaluator.go, line 220 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48720#-M7EkheSDsWMAiokQl6E:-M7dffBP7qprPQxWYELt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4306dcb090fc0bdb224fb49f606cb1b423d6f4aa/pkg/sql/rowexec/inverted_expr_evaluator.go#L220)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

invertedValRoutingInfo -> invertedSpanRoutingInfo
</blockquote></details>

Done.
___
*[pkg/sql/rowexec/inverted_expr_evaluator.go, line 221 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48720#-M7EkrYU6t9GMgl0J92i:-M7dfp3z-YZc9sx432eN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4306dcb090fc0bdb224fb49f606cb1b423d6f4aa/pkg/sql/rowexec/inverted_expr_evaluator.go#L221)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] needs -> need
</blockquote></details>

Done.
___
*[pkg/sql/rowexec/inverted_expr_evaluator.go, line 231 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48720#-M7ElGGp3EBfAs25dAsN:-M7dg2R_D6YmozYRWgwQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4306dcb090fc0bdb224fb49f606cb1b423d6f4aa/pkg/sql/rowexec/inverted_expr_evaluator.go#L231)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

RPInvertedIndexExpr -> *invertedexpr.SpanExpression
</blockquote></details>

Done.
___
*[pkg/sql/rowexec/inverted_expr_evaluator.go, line 265 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48720#-M7En74x2XjP-vdOmRFW:-M7dk615DywOJFM3UeAB:b-j82k3c) ([raw file](https://github.com/cockroachdb/cockroach/blob/4306dcb090fc0bdb224fb49f606cb1b423d6f4aa/pkg/sql/rowexec/inverted_expr_evaluator.go#L265)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Not sure I understand what you mean by "the spans are unioned" in this context. Isn't it just that you append the `exprAndSetIndexList` slice from each matching item in `pendingSpans` to the output `exprAndSetIndexList` slice in `fragmentedSpans`? I think you may be overloading the term "union" here....
</blockquote></details>

Yes, I should not have used "union" -- I've replaced it with "appended".
___
*[pkg/sql/rowexec/inverted_expr_evaluator.go, line 319 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48720#-M7Epcch6Wus7T4ukfKJ:-M7dgDlz6F7mqQEiKZ4A:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4306dcb090fc0bdb224fb49f606cb1b423d6f4aa/pkg/sql/rowexec/inverted_expr_evaluator.go#L319)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

exprAndOperand -> exprAndSetIndexList
</blockquote></details>

Done.
___
*[pkg/sql/rowexec/inverted_expr_evaluator.go, line 326 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48720#-M7EqUZ0BD5MLkDQWXTK:-M7dmG2H31AJnqteaN0T:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4306dcb090fc0bdb224fb49f606cb1b423d6f4aa/pkg/sql/rowexec/inverted_expr_evaluator.go#L326)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I think this could use a comment to remind us why you're breaking here.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> Reviewed 5 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48720)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @sumeerbhola)

___
*[pkg/sql/rowexec/inverted_expr_evaluator_test.go, line 26 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48720#-M7wmQRZ6KAiLcnJ9t-q:-M7wmQRZ6KAiLcnJ9t-r:bx7elo1) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4db05b2cdc57d1c64c555c027a9eb5dcc895c18/pkg/sql/rowexec/inverted_expr_evaluator_test.go#L26)):*
> ```Go
> 	var b strings.Builder
> 	for _, elem := range set {
> 		fmt.Fprintf(&b, "%d,", elem)
> ```

[nit] do you want to skip the comma for the last element?
___
*[pkg/sql/rowexec/inverted_expr_evaluator_test.go, line 81 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48720#-M7wnTTl7ek9F5AgKZUA:-M7wnTTl7ek9F5AgKZUB:bkx6x0m) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4db05b2cdc57d1c64c555c027a9eb5dcc895c18/pkg/sql/rowexec/inverted_expr_evaluator_test.go#L81)):*
> ```Go
> }
> 
> func TestSetContainerUnion(t *testing.T) {
> ```

[nit] it would be easier to follow these tests if you separate the test cases from the code to check for correctness. For example, put the test cases into a `[]testCase{}` and make a function that calls `require.Equal`.
___
*[pkg/sql/rowexec/inverted_expr_evaluator_test.go, line 182 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48720#-M7wq-vm0PDrXn7P1xKg:-M7wq-vnB606pCZ85TfS:b-j6rd9r) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4db05b2cdc57d1c64c555c027a9eb5dcc895c18/pkg/sql/rowexec/inverted_expr_evaluator_test.go#L182)):*
> ```Go
> 	}
> 	expectedSpans := "[a, n) "
> 	expectedFragmentedSpans :=
> ```

nice test!
___
*[pkg/sql/rowexec/inverted_expr_evaluator_test.go, line 203 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48720#-M7wqYINEPcwmT-f0ImO:-M7wqYINEPcwmT-f0ImP:b-mxgle0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4db05b2cdc57d1c64c555c027a9eb5dcc895c18/pkg/sql/rowexec/inverted_expr_evaluator_test.go#L203)):*
> ```Go
> 
> 	// Construct the test input.
> 	indexRows := []keyAndIndex{{"ii", 0}, {"a", 1}, {"ee", 2},
> ```

are these supposed to be rows in an inverted index? Why wouldn't they be ordered on the key?
___
*[pkg/sql/rowexec/inverted_expr_evaluator_test.go, line 214 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48720#-M7wrOF598tSl2XdVyfc:-M7wrOF598tSl2XdVyfd:b-vgoi0o) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4db05b2cdc57d1c64c555c027a9eb5dcc895c18/pkg/sql/rowexec/inverted_expr_evaluator_test.go#L214)):*
> ```Go
> 		batchEvalIntersection.addIndexRow(invertedexpr.EncInvertedVal(elem.key), KeyIndex(elem.index))
> 	}
> 	expectedUnion := "0: 0 3 4 5 6 7 8 \n1: \n"
> ```

I don't understand why this is the expected output (here and the rest of this test case)
___
*[pkg/sql/rowexec/inverted_expr_evaluator_test.go, line 259 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48720#-M7wsYDJCdGia1lgqNZV:-M7wsYDJCdGia1lgqNZW:b-wcu84y) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4db05b2cdc57d1c64c555c027a9eb5dcc895c18/pkg/sql/rowexec/inverted_expr_evaluator_test.go#L259)):*
> ```Go
> 	}
> 	require.Equal(t, "[a, l) [o, p) ", spansToString(batchEval.getSpans()))
> 	require.Equal(t,
> ```

Is it worth adding a test where there are multiple expressions and multiple sets per expression?


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48720)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @rytaft, and @sumeerbhola)

___
*[pkg/sql/rowexec/inverted_expr_evaluator_test.go, line 26 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48720#-M7wmQRZ6KAiLcnJ9t-q:-M7xTJVb5s5EpREdXIQr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4db05b2cdc57d1c64c555c027a9eb5dcc895c18/pkg/sql/rowexec/inverted_expr_evaluator_test.go#L26)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] do you want to skip the comma for the last element?
</blockquote></details>

Done.
___
*[pkg/sql/rowexec/inverted_expr_evaluator_test.go, line 81 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48720#-M7wnTTl7ek9F5AgKZUA:-M7xVrwQ8tmecNTJ7NYj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4db05b2cdc57d1c64c555c027a9eb5dcc895c18/pkg/sql/rowexec/inverted_expr_evaluator_test.go#L81)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] it would be easier to follow these tests if you separate the test cases from the code to check for correctness. For example, put the test cases into a `[]testCase{}` and make a function that calls `require.Equal`.
</blockquote></details>

Done.
___
*[pkg/sql/rowexec/inverted_expr_evaluator_test.go, line 203 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48720#-M7wqYINEPcwmT-f0ImO:-M7xWZKz45yzI_ZOVtnS:b-ebkfbk) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4db05b2cdc57d1c64c555c027a9eb5dcc895c18/pkg/sql/rowexec/inverted_expr_evaluator_test.go#L203)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

are these supposed to be rows in an inverted index? Why wouldn't they be ordered on the key?
</blockquote></details>

Yes, these are the inverted column and the index of the primary key. They are not in order since the evaluator does not care. I've added a comment.
Also, I am guessing we allow scans that span ranges (say the inverted index was large enough to be split across many ranges) to be issued concurrently to those ranges, and so the results could be interleaved and not in sorted order -- is that actually true?
___
*[pkg/sql/rowexec/inverted_expr_evaluator_test.go, line 214 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48720#-M7wrOF598tSl2XdVyfc:-M7xXdGd5HWSM3FQX-oA:b-ctgg7a) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4db05b2cdc57d1c64c555c027a9eb5dcc895c18/pkg/sql/rowexec/inverted_expr_evaluator_test.go#L214)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I don't understand why this is the expected output (here and the rest of this test case)
</blockquote></details>

I hand constructed these, just like the test case (which was designed to exercise a bunch of code paths).
I may also add some randomized cases -- I've added a TODO for a future PR to do that.
___
*[pkg/sql/rowexec/inverted_expr_evaluator_test.go, line 259 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48720#-M7wsYDJCdGia1lgqNZV:-M7xYkiW9HV6p3ZzaLQY:b8nwoxk) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4db05b2cdc57d1c64c555c027a9eb5dcc895c18/pkg/sql/rowexec/inverted_expr_evaluator_test.go#L259)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Is it worth adding a test where there are multiple expressions and multiple sets per expression?
</blockquote></details>

The previous test did have multiple expressions in `batchBoth`, each with multiple sets. And those sets had some complex overlap but not across expressions. This one is trying to test the fragmentation induced by overlaps across expressions.
I looked at test coverage and made a couple of other tweaks and it is looking pretty good
```
github.com/cockroachdb/cockroach/pkg/sql/rowexec/inverted_expr_evaluator.go:49:		unionSetContainers			100.0%
github.com/cockroachdb/cockroach/pkg/sql/rowexec/inverted_expr_evaluator.go:80:		intersectSetContainers			100.0%
github.com/cockroachdb/cockroach/pkg/sql/rowexec/inverted_expr_evaluator.go:131:	newInvertedExprEvaluator		100.0%
github.com/cockroachdb/cockroach/pkg/sql/rowexec/inverted_expr_evaluator.go:137:	initSetExpr				100.0%
github.com/cockroachdb/cockroach/pkg/sql/rowexec/inverted_expr_evaluator.go:158:	getSpansAndSetIndex			100.0%
github.com/cockroachdb/cockroach/pkg/sql/rowexec/inverted_expr_evaluator.go:166:	addIndexRow				100.0%
github.com/cockroachdb/cockroach/pkg/sql/rowexec/inverted_expr_evaluator.go:175:	evaluate				100.0%
github.com/cockroachdb/cockroach/pkg/sql/rowexec/inverted_expr_evaluator.go:195:	evaluateSetExpr				100.0%
github.com/cockroachdb/cockroach/pkg/sql/rowexec/inverted_expr_evaluator.go:279:	fragmentPendingSpans			95.8%
github.com/cockroachdb/cockroach/pkg/sql/rowexec/inverted_expr_evaluator.go:337:	pendingLenWithSameEnd			100.0%
github.com/cockroachdb/cockroach/pkg/sql/rowexec/inverted_expr_evaluator.go:350:	getSpans				97.4%
github.com/cockroachdb/cockroach/pkg/sql/rowexec/inverted_expr_evaluator.go:419:	addIndexRow				100.0%
github.com/cockroachdb/cockroach/pkg/sql/rowexec/inverted_expr_evaluator.go:431:	evaluate				100.0%
github.com/cockroachdb/cockroach/pkg/sql/rowexec/inverted_expr_evaluator.go:442:	reset					100.0%
```


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> Reviewed 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48720)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @sumeerbhola)

___
*[pkg/sql/rowexec/inverted_expr_evaluator_test.go, line 203 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48720#-M7wqYINEPcwmT-f0ImO:-M8GSg0MAvdkgk1Q4CF6:b-6tm0h9) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4db05b2cdc57d1c64c555c027a9eb5dcc895c18/pkg/sql/rowexec/inverted_expr_evaluator_test.go#L203)):*
> so the results could be interleaved and not in sorted order -- is that actually true?

I'm not sure, but that sounds right
___
*[pkg/sql/rowexec/inverted_expr_evaluator_test.go, line 234 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48720#-M8GTrA-BuCwl55OMRFv:-M8GTrA-BuCwl55OMRFw:b-r8ozd4) ([raw file](https://github.com/cockroachdb/cockroach/blob/801bc48e2a4749975d6b95624ba032efa3082836/pkg/sql/rowexec/inverted_expr_evaluator_test.go#L234)):*
> ```Go
> 	// These expected values were hand-constructed by evaluating the expression over
> 	// these inputs.
> 	expectedUnion := "0: 0 3 4 5 6 7 8 \n1: \n"
> ```

I think it might help to add a reminder in the comment of what the expression looks like so we can visually see why this is the right output.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48720)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @rytaft, and @sumeerbhola)

___
*[pkg/sql/rowexec/inverted_expr_evaluator_test.go, line 203 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48720#-M7wqYINEPcwmT-f0ImO:-M8HeoB74jhbkCZUnKGk:b-aiu5yn) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4db05b2cdc57d1c64c555c027a9eb5dcc895c18/pkg/sql/rowexec/inverted_expr_evaluator_test.go#L203)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

> so the results could be interleaved and not in sorted order -- is that actually true?

I'm not sure, but that sounds right
</blockquote></details>

According to @asubiotto they will be in sorted order since the index is scanned sequentially.
But the evaluator does not care, yet. There is a TODO in `addIndexRow` to use this ordering to optimize the binary search -- this test will change when that TODO is done (later PR).
___
*[pkg/sql/rowexec/inverted_expr_evaluator_test.go, line 234 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48720#-M8GTrA-BuCwl55OMRFv:-M8Hq_kEAXYrhQ0-Wam9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/801bc48e2a4749975d6b95624ba032efa3082836/pkg/sql/rowexec/inverted_expr_evaluator_test.go#L234)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I think it might help to add a reminder in the comment of what the expression looks like so we can visually see why this is the right output.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48720#-:-M8HyVoc3-Ck2MpO4BiU:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto and @sumeerbhola)



<!-- Sent from Reviewable.io -->

sumeerbhola commented:
bors r+
